Error on missing vpn.outDir

`dockerOpenVPN.generateCerts` fails with the following error on OSX:

```
panic: couldn't deploy the vpn: Error response from daemon: invalid mount config for type "bind": bind source path does not exist: /Users/<user>/.config/simnet/vpn
```

We should either create create the directory in `NewOptions` or update the Readme with a remark about this requirement.

--- a/daemon/monitor/netem/executor.go
+++ b/daemon/monitor/netem/executor.go
@@ -54,7 +54,7 @@ func (e executor) Execute(rules []network.Rule) error {
 	for _, args := range commands {
 		err := e.execTc(args)
 		if err != nil {
-			return fmt.Errorf("%s command failed: %v", e.cmd, err)
+			return xerrors.Errorf("%s command failed: %v", e.cmd, err)
 		}
 	}
 
@@ -70,7 +70,7 @@ func (e executor) execTc(args []string) error {
 
 	err := cmd.Run()
 	if err != nil {
-		return fmt.Errorf("cmd failed: %s", err)
+		return xerrors.Errorf("cmd failed: %s", err)
 	}
 
 	return nil

--- a/daemon/monitor/netem/executor.go
+++ b/daemon/monitor/netem/executor.go
@@ -7,6 +7,7 @@ import (
 	"strings"
 
 	"go.dedis.ch/simnet/network"
+	"golang.org/x/xerrors"
 )
 
 // Executor is responsible for executing the commands necessary to apply the
@@ -70,7 +71,7 @@ func (e executor) execTc(args []string) error {
 
 	err := cmd.Run()
 	if err != nil {
-		return xerrors.Errorf("cmd failed: %s", err)
+		return xerrors.Errorf("cmd failed: %v", err)
 	}
 
 	return nil

--- a/daemon/monitor/netem/main.go
+++ b/daemon/monitor/netem/main.go
@@ -23,7 +23,7 @@ const (
 
 func checkErr(err error, msg string) {
 	if err != nil {
-		panic(fmt.Errorf("%s: %v", msg, err))
+		panic(xerrors.Errorf("%s: %v", msg, err))
 	}
 }
 

--- a/daemon/monitor/netem/main.go
+++ b/daemon/monitor/netem/main.go
@@ -7,6 +7,7 @@ import (
 	"os"
 
 	"go.dedis.ch/simnet/network"
+	"golang.org/x/xerrors"
 )
 
 const (

--- a/mod.go
+++ b/mod.go
@@ -79,24 +79,19 @@ func (s *Simulation) Run(args []string) error {
 	if doDeploy || doAll {
 		err := s.strategy.Deploy(ctx, s.round)
 		if err != nil {
-			return err
+			return xerrors.Errorf("couldn't deploy experiment: %w", err)
 		}
 	}
 
 	if doExecute || doAll {
 		err := s.strategy.Execute(ctx, s.round)
 		if err != nil {
-			return err
-		}
-
-		err = s.strategy.WriteStats(ctx, "result.json")
-		if err != nil {
-			return err
+			return xerrors.Errorf("couldn't execute experiment: %w", err)
 		}
 	}
 
-	if doStats {
-		fmt.Fprintln(s.out, "Write statistics only")
+	if doStats || doExecute || doAll {
+		fmt.Fprintln(s.out, "Writing statistics")
 
 		// The user can request to only write the statistics to the file.
 		err := s.strategy.WriteStats(ctx, "result.json")

--- a/mod.go
+++ b/mod.go
@@ -66,9 +66,11 @@ func (s *Simulation) Run(args []string) error {
 			// Anything bad happening during the cleaning phase will be printed
 			// so an error of the simulation is returned if any.
 
+			fmt.Println("Starting cleaning...")
+
 			err := s.strategy.Clean(ctx)
 			if err != nil {
-				fmt.Fprintf(s.out, "An error occured during cleaning: %v\n", err)
+				fmt.Fprintln(s.out, "An error occurred during cleaning: ", err)
 				fmt.Fprintln(s.out, "Please make sure to clean remaining components.")
 			}
 		}()

--- a/mod.go
+++ b/mod.go
@@ -79,14 +79,14 @@ func (s *Simulation) Run(args []string) error {
 	if doDeploy || doAll {
 		err := s.strategy.Deploy(ctx, s.round)
 		if err != nil {
-			return xerrors.Errorf("couldn't deploy experiment: %w", err)
+			return xerrors.Errorf("couldn't deploy experiment: %v", err)
 		}
 	}
 
 	if doExecute || doAll {
 		err := s.strategy.Execute(ctx, s.round)
 		if err != nil {
-			return xerrors.Errorf("couldn't execute experiment: %w", err)
+			return xerrors.Errorf("couldn't execute experiment: %v", err)
 		}
 	}
 

--- a/mod_test.go
+++ b/mod_test.go
@@ -100,7 +100,7 @@ func TestSimulation_Run(t *testing.T) {
 	stry.errClean = errors.New("clean")
 	err = sim.Run(args)
 	require.NoError(t, err)
-	require.Contains(t, buffer.String(), "An error occured during cleaning")
+	require.Contains(t, buffer.String(), "An error occurred during cleaning")
 
 	args = []string{os.Args[0], "--do-stats"}
 	stry.errStats = errors.New("oops")

--- a/sim/docker/io.go
+++ b/sim/docker/io.go
@@ -76,7 +76,7 @@ func (dio *dockerio) Write(container, path string, content io.Reader) error {
 	buffer := new(bytes.Buffer)
 	_, err := io.Copy(buffer, content)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed copying data into buffer: %w", err)
 	}
 
 	go func() {
@@ -90,7 +90,7 @@ func (dio *dockerio) Write(container, path string, content io.Reader) error {
 
 	err = dio.cli.CopyToContainer(ctx, container, path, reader, types.CopyToContainerOptions{})
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed copying buffer into container: %w", err)
 	}
 
 	return nil

--- a/sim/docker/io.go
+++ b/sim/docker/io.go
@@ -5,7 +5,6 @@ import (
 	"bytes"
 	"context"
 	"encoding/json"
-	"fmt"
 	"io"
 	"io/ioutil"
 	"math"
@@ -52,12 +51,12 @@ func (dio *dockerio) Read(container, path string) (io.ReadCloser, error) {
 
 	reader, _, err := dio.cli.CopyFromContainer(ctx, container, path)
 	if err != nil {
-		return nil, fmt.Errorf("couldn't open stream: %w", err)
+		return nil, xerrors.Errorf("couldn't open stream: %v", err)
 	}
 
 	tr := tar.NewReader(reader)
 	if _, err = tr.Next(); err != nil {
-		return nil, fmt.Errorf("couldn't untar: %w", err)
+		return nil, xerrors.Errorf("couldn't untar: %v", err)
 	}
 
 	return ioutil.NopCloser(tr), nil
@@ -76,7 +75,7 @@ func (dio *dockerio) Write(container, path string, content io.Reader) error {
 	buffer := new(bytes.Buffer)
 	_, err := io.Copy(buffer, content)
 	if err != nil {
-		return xerrors.Errorf("failed copying data into buffer: %w", err)
+		return xerrors.Errorf("failed copying data into buffer: %v", err)
 	}
 
 	go func() {
@@ -90,7 +89,7 @@ func (dio *dockerio) Write(container, path string, content io.Reader) error {
 
 	err = dio.cli.CopyToContainer(ctx, container, path, reader, types.CopyToContainerOptions{})
 	if err != nil {
-		return xerrors.Errorf("failed copying buffer into container: %w", err)
+		return xerrors.Errorf("failed copying buffer into container: %v", err)
 	}
 
 	return nil
@@ -108,14 +107,14 @@ func (dio *dockerio) Exec(container string, cmd []string, options sim.ExecOption
 		Cmd:          cmd,
 	})
 	if err != nil {
-		return fmt.Errorf("couldn't create exec: %w", err)
+		return xerrors.Errorf("couldn't create exec: %v", err)
 	}
 
 	msgCh, errCh := dio.cli.Events(ctx, types.EventsOptions{})
 
 	conn, err := dio.cli.ContainerExecAttach(ctx, resp.ID, types.ExecConfig{})
 	if err != nil {
-		return fmt.Errorf("couldn't attach exec: %w", err)
+		return xerrors.Errorf("couldn't attach exec: %v", err)
 	}
 
 	defer conn.Close()
@@ -153,13 +152,13 @@ func (dio *dockerio) Exec(container string, cmd []string, options sim.ExecOption
 
 	err = dio.cli.ContainerExecStart(ctx, resp.ID, types.ExecStartCheck{})
 	if err != nil {
-		return fmt.Errorf("couldn't start exec: %w", err)
+		return xerrors.Errorf("couldn't start exec: %v", err)
 	}
 
 	// Check if something bad happened when writing to stdin.
 	err, ok := <-done
 	if err != nil && ok {
-		return fmt.Errorf("couldn't write to stdin: %v", err)
+		return xerrors.Errorf("couldn't write to stdin: %v", err)
 	}
 
 	// Everything's good so far so let's wait for the command to be completly done
@@ -170,13 +169,13 @@ func (dio *dockerio) Exec(container string, cmd []string, options sim.ExecOption
 			if msg.Status == "exec_die" && msg.Actor.Attributes["execID"] == resp.ID {
 				code := msg.Actor.Attributes["exitCode"]
 				if code != "0" {
-					return fmt.Errorf("exited with %s", code)
+					return xerrors.Errorf("exited with %s", code)
 				}
 
 				return nil
 			}
 		case err := <-errCh:
-			return fmt.Errorf("received error event: %w", err)
+			return xerrors.Errorf("received error event: %w", err)
 		}
 	}
 }
@@ -210,7 +209,7 @@ func (dio *dockerio) FetchStats(from, end time.Time, filename string) error {
 	enc := json.NewEncoder(file)
 	err = enc.Encode(&dio.stats)
 	if err != nil {
-		return xerrors.Errorf("couldn't encode the stats: %w", err)
+		return xerrors.Errorf("couldn't encode the stats: %v", err)
 	}
 
 	return nil

--- a/sim/docker/mod.go
+++ b/sim/docker/mod.go
@@ -168,7 +168,7 @@ func pullImage(ctx context.Context, cli client.APIClient, image string, out io.W
 
 	reader, err := cli.ImagePull(ctx, ref, types.ImagePullOptions{})
 	if err != nil {
-		return xerrors.Errorf("failed pulling image '%v': %v", ref, err)
+		return xerrors.Errorf("failed pulling image '%s': %v", ref, err)
 	}
 
 	defer reader.Close()

--- a/sim/docker/mod.go
+++ b/sim/docker/mod.go
@@ -130,7 +130,7 @@ func (s *Strategy) refreshContainers(ctx context.Context) error {
 		Filters: args,
 	})
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed refreshing containers: %w", err)
 	}
 
 	s.containers = containers
@@ -168,14 +168,14 @@ func pullImage(ctx context.Context, cli client.APIClient, image string, out io.W
 
 	reader, err := cli.ImagePull(ctx, ref, types.ImagePullOptions{})
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed pulling image '%v': %w", ref, err)
 	}
 
 	defer reader.Close()
 
 	err = waitImagePull(reader, image, out)
 	if err != nil {
-		return fmt.Errorf("couldn't complete pull: %w", err)
+		return xerrors.Errorf("couldn't complete pull: %w", err)
 	}
 
 	return nil
@@ -215,18 +215,18 @@ func (s *Strategy) createContainers(ctx context.Context) error {
 
 		resp, err := s.cli.ContainerCreate(ctx, cfg, hcfg, nil, node.String())
 		if err != nil {
-			return err
+			return xerrors.Errorf("failed creating container: %w", err)
 		}
 
 		err = s.cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{})
 		if err != nil {
-			return err
+			return xerrors.Errorf("failed starting container: %w", err)
 		}
 	}
 
 	err := s.refreshContainers(ctx)
 	if err != nil {
-		return fmt.Errorf("couldn't refresh the list of containers: %w", err)
+		return xerrors.Errorf("couldn't refresh the list of containers: %w", err)
 	}
 
 	return nil
@@ -280,7 +280,7 @@ func (s *Strategy) configureContainer(
 		Stderr: true,
 	})
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed attaching container: %w", err)
 	}
 
 	defer conn.Close()
@@ -349,7 +349,7 @@ func (s *Strategy) configureContainers(ctx context.Context) error {
 		err = s.configureContainer(ctx, c, cfg, mapping, msgCh, errCh)
 		if err != nil {
 			fmt.Fprintln(s.out, goterm.ResetLine("Configure containers... Failed."))
-			return err
+			return xerrors.Errorf("failed configuring container: %w", err)
 		}
 	}
 
@@ -375,7 +375,7 @@ func (s *Strategy) Deploy(ctx context.Context, round sim.Round) error {
 
 	err = s.streamLogs(ctx)
 	if err != nil {
-		return xerrors.Errorf("couldn't stream logs: %v", err)
+		return xerrors.Errorf("couldn't stream logs: %w", err)
 	}
 
 	err = s.configureContainers(ctx)
@@ -385,12 +385,12 @@ func (s *Strategy) Deploy(ctx context.Context, round sim.Round) error {
 
 	err = s.vpn.Deploy()
 	if err != nil {
-		return xerrors.Errorf("couldn't deploy the vpn: %v", err)
+		return xerrors.Errorf("couldn't deploy the vpn: %w", err)
 	}
 
 	err = round.Before(s.dio, s.makeExecutionContext())
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed running 'Before': %w", err)
 	}
 
 	fmt.Fprintln(s.out, "Deployment... Done.")
@@ -422,12 +422,12 @@ func (s *Strategy) streamLogs(ctx context.Context) error {
 	// Clean the folder so it does not mix different simulations.
 	err := os.RemoveAll(logFolder)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed removing logs: %w", err)
 	}
 
 	err = os.MkdirAll(logFolder, 0755)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed creating log directory: %w", err)
 	}
 
 	for _, container := range s.containers {
@@ -439,12 +439,12 @@ func (s *Strategy) streamLogs(ctx context.Context) error {
 		})
 
 		if err != nil {
-			return err
+			return xerrors.Errorf("failed accessing container logs: %w", err)
 		}
 
 		f, err := os.Create(filepath.Join(logFolder, container.Names[0]))
 		if err != nil {
-			return err
+			return xerrors.Errorf("failed creating log folder: %w", err)
 		}
 
 		go func() {
@@ -461,12 +461,12 @@ func (s *Strategy) streamLogs(ctx context.Context) error {
 func (s *Strategy) Execute(ctx context.Context, round sim.Round) error {
 	err := s.refreshContainers(ctx)
 	if err != nil {
-		return fmt.Errorf("couldn't update the states: %w", err)
+		return xerrors.Errorf("couldn't update the states: %w", err)
 	}
 
 	closer, err := s.dio.monitorContainers(ctx, s.containers)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed monitoring containers: %w", err)
 	}
 
 	defer closer()
@@ -475,7 +475,7 @@ func (s *Strategy) Execute(ctx context.Context, round sim.Round) error {
 	// for each of them in the output folder.
 	err = s.streamLogs(ctx)
 	if err != nil {
-		return xerrors.Errorf("couldn't stream logs: %v", err)
+		return xerrors.Errorf("failed streaming logs: %w", err)
 	}
 
 	nodes := s.makeExecutionContext()
@@ -502,7 +502,7 @@ func (s *Strategy) WriteStats(ctx context.Context, filename string) error {
 	out := filepath.Join(s.options.OutputDir, filename)
 	err := s.dio.FetchStats(time.Unix(0, 0), time.Now(), out)
 	if err != nil {
-		return xerrors.Errorf("couldn't fetch stats: %v", err)
+		return xerrors.Errorf("failed fetching stats: %w", err)
 	}
 
 	fmt.Fprintln(s.out, "Write statistics... Done.")
@@ -516,7 +516,7 @@ func (s *Strategy) Clean(ctx context.Context) error {
 
 	err := s.vpn.Clean()
 	if err != nil {
-		errs = append(errs, fmt.Errorf("vpn: %v", err))
+		errs = append(errs, xerrors.Errorf("vpn: %w", err))
 	}
 
 	err = s.refreshContainers(ctx)

--- a/sim/docker/mod.go
+++ b/sim/docker/mod.go
@@ -130,7 +130,7 @@ func (s *Strategy) refreshContainers(ctx context.Context) error {
 		Filters: args,
 	})
 	if err != nil {
-		return xerrors.Errorf("failed refreshing containers: %w", err)
+		return xerrors.Errorf("failed refreshing containers: %v", err)
 	}
 
 	s.containers = containers
@@ -151,12 +151,12 @@ func waitImagePull(reader io.Reader, image string, out io.Writer) error {
 			}
 
 			fmt.Fprintln(out, goterm.ResetLine("Pull image... Failed."))
-			return fmt.Errorf("couldn't decode the event: %w", err)
+			return xerrors.Errorf("couldn't decode the event: %v", err)
 		}
 
 		if evt.Error != "" {
 			// Typically a stream errors or a server side error.
-			return fmt.Errorf("stream error: %s", evt.Error)
+			return xerrors.Errorf("stream error: %s", evt.Error)
 		}
 
 		fmt.Fprintf(out, goterm.ResetLine("Pull image... %s %s"), evt.Status, evt.Progress)
@@ -168,14 +168,14 @@ func pullImage(ctx context.Context, cli client.APIClient, image string, out io.W
 
 	reader, err := cli.ImagePull(ctx, ref, types.ImagePullOptions{})
 	if err != nil {
-		return xerrors.Errorf("failed pulling image '%v': %w", ref, err)
+		return xerrors.Errorf("failed pulling image '%v': %v", ref, err)
 	}
 
 	defer reader.Close()
 
 	err = waitImagePull(reader, image, out)
 	if err != nil {
-		return xerrors.Errorf("couldn't complete pull: %w", err)
+		return xerrors.Errorf("couldn't complete pull: %v", err)
 	}
 
 	return nil
@@ -215,18 +215,18 @@ func (s *Strategy) createContainers(ctx context.Context) error {
 
 		resp, err := s.cli.ContainerCreate(ctx, cfg, hcfg, nil, node.String())
 		if err != nil {
-			return xerrors.Errorf("failed creating container: %w", err)
+			return xerrors.Errorf("failed creating container: %v", err)
 		}
 
 		err = s.cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{})
 		if err != nil {
-			return xerrors.Errorf("failed starting container: %w", err)
+			return xerrors.Errorf("failed starting container: %v", err)
 		}
 	}
 
 	err := s.refreshContainers(ctx)
 	if err != nil {
-		return xerrors.Errorf("couldn't refresh the list of containers: %w", err)
+		return xerrors.Errorf("couldn't refresh the list of containers: %v", err)
 	}
 
 	return nil
@@ -241,7 +241,7 @@ func waitExec(containerID string, msgCh <-chan events.Message, errCh <-chan erro
 			if msg.Status == "die" && msg.Actor.ID == containerID {
 				code := msg.Actor.Attributes["exitCode"]
 				if code != "0" {
-					return fmt.Errorf("exit code %s", code)
+					return xerrors.Errorf("exit code %s", code)
 				}
 
 				return nil
@@ -270,7 +270,7 @@ func (s *Strategy) configureContainer(
 
 	resp, err := s.cli.ContainerCreate(ctx, cfg, hcfg, nil, "")
 	if err != nil {
-		return fmt.Errorf("couldn't create netem container: %w", err)
+		return xerrors.Errorf("couldn't create netem container: %v", err)
 	}
 
 	conn, err := s.cli.ContainerAttach(ctx, resp.ID, types.ContainerAttachOptions{
@@ -280,14 +280,14 @@ func (s *Strategy) configureContainer(
 		Stderr: true,
 	})
 	if err != nil {
-		return xerrors.Errorf("failed attaching container: %w", err)
+		return xerrors.Errorf("failed attaching container: %v", err)
 	}
 
 	defer conn.Close()
 
 	err = s.cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{})
 	if err != nil {
-		return fmt.Errorf("couldn't start netem container: %w", err)
+		return xerrors.Errorf("couldn't start netem container: %v", err)
 	}
 
 	rules := s.options.Topology.Rules(network.NodeID(containerName(c)), mapping)
@@ -295,7 +295,7 @@ func (s *Strategy) configureContainer(
 	enc := json.NewEncoder(conn.Conn)
 	err = enc.Encode(&rules)
 	if err != nil {
-		return fmt.Errorf("couldn't encode the rules: %w", err)
+		return xerrors.Errorf("couldn't encode the rules: %v", err)
 	}
 
 	// Output from the monitor container executing the netem tool.
@@ -303,7 +303,7 @@ func (s *Strategy) configureContainer(
 
 	err = waitExec(resp.ID, msgCh, errCh, ExecWaitTimeout)
 	if err != nil {
-		return fmt.Errorf("couldn't apply rule: %w", err)
+		return xerrors.Errorf("couldn't apply rule: %v", err)
 	}
 
 	return nil
@@ -313,14 +313,14 @@ func (s *Strategy) configureContainers(ctx context.Context) error {
 	ref := fmt.Sprintf("%s/%s:%s", ImageBaseURL, ImageMonitor, daemon.Version)
 	reader, err := s.cli.ImagePull(ctx, ref, types.ImagePullOptions{})
 	if err != nil {
-		return fmt.Errorf("couldn't pull netem image: %w", err)
+		return xerrors.Errorf("couldn't pull netem image: %v", err)
 	}
 
 	defer reader.Close()
 
 	err = waitImagePull(reader, ref, s.out)
 	if err != nil {
-		return fmt.Errorf("couldn't complete pull: %w", err)
+		return xerrors.Errorf("couldn't complete pull: %v", err)
 	}
 
 	cfg := &container.Config{
@@ -349,7 +349,7 @@ func (s *Strategy) configureContainers(ctx context.Context) error {
 		err = s.configureContainer(ctx, c, cfg, mapping, msgCh, errCh)
 		if err != nil {
 			fmt.Fprintln(s.out, goterm.ResetLine("Configure containers... Failed."))
-			return xerrors.Errorf("failed configuring container: %w", err)
+			return xerrors.Errorf("failed configuring container: %v", err)
 		}
 	}
 
@@ -362,35 +362,35 @@ func (s *Strategy) configureContainers(ctx context.Context) error {
 func (s *Strategy) Deploy(ctx context.Context, round sim.Round) error {
 	err := pullImage(ctx, s.cli, s.options.Image, s.out)
 	if err != nil {
-		return xerrors.Errorf("couldn't pull the image: %w", err)
+		return xerrors.Errorf("couldn't pull the image: %v", err)
 	}
 
 	fmt.Fprintf(s.out, "Creating containers... In Progress.")
 	err = s.createContainers(ctx)
 	if err != nil {
 		fmt.Fprintln(s.out, goterm.ResetLine("Creating containers... Failed."))
-		return xerrors.Errorf("couldn't create the container: %w", err)
+		return xerrors.Errorf("couldn't create the container: %v", err)
 	}
 	fmt.Fprintln(s.out, goterm.ResetLine("Creating containers... Done."))
 
 	err = s.streamLogs(ctx)
 	if err != nil {
-		return xerrors.Errorf("couldn't stream logs: %w", err)
+		return xerrors.Errorf("couldn't stream logs: %v", err)
 	}
 
 	err = s.configureContainers(ctx)
 	if err != nil {
-		return xerrors.Errorf("couldn't configure the containers: %w", err)
+		return xerrors.Errorf("couldn't configure the containers: %v", err)
 	}
 
 	err = s.vpn.Deploy()
 	if err != nil {
-		return xerrors.Errorf("couldn't deploy the vpn: %w", err)
+		return xerrors.Errorf("couldn't deploy the vpn: %v", err)
 	}
 
 	err = round.Before(s.dio, s.makeExecutionContext())
 	if err != nil {
-		return xerrors.Errorf("failed running 'Before': %w", err)
+		return xerrors.Errorf("failed running 'Before': %v", err)
 	}
 
 	fmt.Fprintln(s.out, "Deployment... Done.")
@@ -422,12 +422,12 @@ func (s *Strategy) streamLogs(ctx context.Context) error {
 	// Clean the folder so it does not mix different simulations.
 	err := os.RemoveAll(logFolder)
 	if err != nil {
-		return xerrors.Errorf("failed removing logs: %w", err)
+		return xerrors.Errorf("failed removing logs: %v", err)
 	}
 
 	err = os.MkdirAll(logFolder, 0755)
 	if err != nil {
-		return xerrors.Errorf("failed creating log directory: %w", err)
+		return xerrors.Errorf("failed creating log directory: %v", err)
 	}
 
 	for _, container := range s.containers {
@@ -439,12 +439,12 @@ func (s *Strategy) streamLogs(ctx context.Context) error {
 		})
 
 		if err != nil {
-			return xerrors.Errorf("failed accessing container logs: %w", err)
+			return xerrors.Errorf("failed accessing container logs: %v", err)
 		}
 
 		f, err := os.Create(filepath.Join(logFolder, container.Names[0]))
 		if err != nil {
-			return xerrors.Errorf("failed creating log folder: %w", err)
+			return xerrors.Errorf("failed creating log folder: %v", err)
 		}
 
 		go func() {
@@ -461,12 +461,12 @@ func (s *Strategy) streamLogs(ctx context.Context) error {
 func (s *Strategy) Execute(ctx context.Context, round sim.Round) error {
 	err := s.refreshContainers(ctx)
 	if err != nil {
-		return xerrors.Errorf("couldn't update the states: %w", err)
+		return xerrors.Errorf("couldn't update the states: %v", err)
 	}
 
 	closer, err := s.dio.monitorContainers(ctx, s.containers)
 	if err != nil {
-		return xerrors.Errorf("failed monitoring containers: %w", err)
+		return xerrors.Errorf("failed monitoring containers: %v", err)
 	}
 
 	defer closer()
@@ -475,20 +475,20 @@ func (s *Strategy) Execute(ctx context.Context, round sim.Round) error {
 	// for each of them in the output folder.
 	err = s.streamLogs(ctx)
 	if err != nil {
-		return xerrors.Errorf("failed streaming logs: %w", err)
+		return xerrors.Errorf("failed streaming logs: %v", err)
 	}
 
 	nodes := s.makeExecutionContext()
 
 	err = round.Execute(s.dio, nodes)
 	if err != nil {
-		return fmt.Errorf("couldn't execute: %w", err)
+		return xerrors.Errorf("couldn't execute: %v", err)
 	}
 
 	// After step so that it is executed after each experiment.
 	err = round.After(s.dio, nodes)
 	if err != nil {
-		return fmt.Errorf("couldn't perform after step: %w", err)
+		return xerrors.Errorf("couldn't perform after step: %v", err)
 	}
 
 	fmt.Fprintln(s.out, "Execution... Done.")
@@ -502,7 +502,7 @@ func (s *Strategy) WriteStats(ctx context.Context, filename string) error {
 	out := filepath.Join(s.options.OutputDir, filename)
 	err := s.dio.FetchStats(time.Unix(0, 0), time.Now(), out)
 	if err != nil {
-		return xerrors.Errorf("failed fetching stats: %w", err)
+		return xerrors.Errorf("failed fetching stats: %v", err)
 	}
 
 	fmt.Fprintln(s.out, "Write statistics... Done.")
@@ -516,12 +516,12 @@ func (s *Strategy) Clean(ctx context.Context) error {
 
 	err := s.vpn.Clean()
 	if err != nil {
-		errs = append(errs, xerrors.Errorf("vpn: %w", err))
+		errs = append(errs, xerrors.Errorf("vpn: %v", err))
 	}
 
 	err = s.refreshContainers(ctx)
 	if err != nil {
-		return fmt.Errorf("couldn't get running containers: %w", err)
+		return xerrors.Errorf("couldn't get running containers: %v", err)
 	}
 
 	timeout := ContainerStopTimeout
@@ -534,7 +534,7 @@ func (s *Strategy) Clean(ctx context.Context) error {
 	}
 
 	if len(errs) > 0 {
-		return fmt.Errorf("couldn't remove the containers: %v", errs)
+		return xerrors.Errorf("couldn't remove the containers: %v", errs)
 	}
 
 	fmt.Fprintln(s.out, "Cleaning... Done.")

--- a/sim/docker/vpn_darwin.go
+++ b/sim/docker/vpn_darwin.go
@@ -58,31 +58,31 @@ func (vpn dockerOpenVPN) Deploy() error {
 	// Get the image for the router container.
 	err := pullImage(ctx, vpn.cli, simnetRouterImage, vpn.out)
 	if err != nil {
-		return xerrors.Errorf("failed pulling router image: %w", err)
+		return xerrors.Errorf("failed pulling router image: %v", err)
 	}
 
 	// Get the image for the initialization container.
 	err = pullImage(ctx, vpn.cli, simnetRouterInitImage, vpn.out)
 	if err != nil {
-		return xerrors.Errorf("failed pulling router init image: %w", err)
+		return xerrors.Errorf("failed pulling router init image: %v", err)
 	}
 
 	// Run the init container to generate the certificates.
 	err = vpn.generateCerts(ctx)
 	if err != nil {
-		return xerrors.Errorf("failed generating certs: %w", err)
+		return xerrors.Errorf("failed generating certs: %v", err)
 	}
 
 	// Deploy the router.
 	err = vpn.createRouterContainer(ctx)
 	if err != nil {
-		return xerrors.Errorf("failed creating router container: %w", err)
+		return xerrors.Errorf("failed creating router container: %v", err)
 	}
 
 	// Open a connection to the VPN.
 	err = vpn.connect()
 	if err != nil {
-		return xerrors.Errorf("failed connecting over VPN: %w", err)
+		return xerrors.Errorf("failed connecting over VPN: %v", err)
 	}
 
 	return nil
@@ -106,22 +106,22 @@ func (vpn dockerOpenVPN) generateCerts(ctx context.Context) error {
 
 	err := os.Mkdir(vpn.outDir, 0755)
 	if err != nil && !errors.Is(err, fs.ErrExist) {
-		return xerrors.Errorf("failed creating VPN dir: %w", err)
+		return xerrors.Errorf("failed creating VPN dir: %v", err)
 	}
 
 	resp, err := vpn.cli.ContainerCreate(ctx, cfg, hcfg, nil, initContainerName)
 	if err != nil {
-		return xerrors.Errorf("failed creating router init container: %w", err)
+		return xerrors.Errorf("failed creating router init container: %v", err)
 	}
 
 	err = vpn.cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{})
 	if err != nil {
-		return xerrors.Errorf("failed starting router init container: %w", err)
+		return xerrors.Errorf("failed starting router init container: %v", err)
 	}
 
 	_, err = vpn.cli.ContainerWait(ctx, initContainerName)
 	if err != nil {
-		return xerrors.Errorf("failed waiting for router init container: %w",
+		return xerrors.Errorf("failed waiting for router init container: %v",
 			err)
 	}
 
@@ -154,12 +154,12 @@ func (vpn dockerOpenVPN) createRouterContainer(ctx context.Context) error {
 
 	resp, err := vpn.cli.ContainerCreate(ctx, cfg, hcfg, nil, routerContainerName)
 	if err != nil {
-		return xerrors.Errorf("failed creating router container: %w", err)
+		return xerrors.Errorf("failed creating router container: %v", err)
 	}
 
 	err = vpn.cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{})
 	if err != nil {
-		return xerrors.Errorf("failed starting router container: %w", err)
+		return xerrors.Errorf("failed starting router container: %v", err)
 	}
 
 	return nil
@@ -179,7 +179,7 @@ func (vpn dockerOpenVPN) connect() error {
 		sim.WithCommand(vpn.options.VPNExecutable),
 	)
 	if err != nil {
-		return xerrors.Errorf("failed starting VPN tunnel: %w", err)
+		return xerrors.Errorf("failed starting VPN tunnel: %v", err)
 	}
 
 	return nil

--- a/sim/docker/vpn_darwin.go
+++ b/sim/docker/vpn_darwin.go
@@ -1,10 +1,14 @@
+//go:build darwin
 // +build darwin
 
 package docker
 
 import (
 	"context"
+	"errors"
 	"io"
+	"io/fs"
+	"os"
 	"path/filepath"
 	"time"
 
@@ -15,6 +19,7 @@ import (
 	"github.com/docker/docker/client"
 	"github.com/docker/go-connections/nat"
 	"go.dedis.ch/simnet/sim"
+	"golang.org/x/xerrors"
 )
 
 const (
@@ -53,31 +58,31 @@ func (vpn dockerOpenVPN) Deploy() error {
 	// Get the image for the router container.
 	err := pullImage(ctx, vpn.cli, simnetRouterImage, vpn.out)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed pulling router image: %w", err)
 	}
 
 	// Get the image for the initialization container.
 	err = pullImage(ctx, vpn.cli, simnetRouterInitImage, vpn.out)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed pulling router init image: %w", err)
 	}
 
 	// Run the init container to generate the certificates.
 	err = vpn.generateCerts(ctx)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed generating certs: %w", err)
 	}
 
 	// Deploy the router.
 	err = vpn.createRouterContainer(ctx)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed creating router container: %w", err)
 	}
 
 	// Open a connection to the VPN.
 	err = vpn.connect()
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed connecting over VPN: %w", err)
 	}
 
 	return nil
@@ -99,19 +104,25 @@ func (vpn dockerOpenVPN) generateCerts(ctx context.Context) error {
 		Image: simnetRouterInitImage,
 	}
 
+	err := os.Mkdir(vpn.outDir, 0755)
+	if err != nil && !errors.Is(err, fs.ErrExist) {
+		return xerrors.Errorf("failed creating VPN dir: %w", err)
+	}
+
 	resp, err := vpn.cli.ContainerCreate(ctx, cfg, hcfg, nil, initContainerName)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed creating router init container: %w", err)
 	}
 
 	err = vpn.cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{})
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed starting router init container: %w", err)
 	}
 
 	_, err = vpn.cli.ContainerWait(ctx, initContainerName)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed waiting for router init container: %w",
+			err)
 	}
 
 	return nil
@@ -143,12 +154,12 @@ func (vpn dockerOpenVPN) createRouterContainer(ctx context.Context) error {
 
 	resp, err := vpn.cli.ContainerCreate(ctx, cfg, hcfg, nil, routerContainerName)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed creating router container: %w", err)
 	}
 
 	err = vpn.cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{})
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed starting router container: %w", err)
 	}
 
 	return nil
@@ -168,7 +179,7 @@ func (vpn dockerOpenVPN) connect() error {
 		sim.WithCommand(vpn.options.VPNExecutable),
 	)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed starting VPN tunnel: %w", err)
 	}
 
 	return nil

--- a/sim/kubernetes/engine.go
+++ b/sim/kubernetes/engine.go
@@ -137,12 +137,12 @@ type kubeEngine struct {
 func newKubeEngine(config *rest.Config, ns string, options *sim.Options) (*kubeEngine, error) {
 	client, err := newClient(config)
 	if err != nil {
-		return nil, fmt.Errorf("client: %v", err)
+		return nil, xerrors.Errorf("client: %v", err)
 	}
 
 	u, err := url.Parse(config.Host)
 	if err != nil {
-		return nil, fmt.Errorf("couldn't parse the host: %v", err)
+		return nil, xerrors.Errorf("couldn't parse the host: %v", err)
 	}
 
 	if options.Data[OptionMemoryAlloc] == nil {
@@ -231,7 +231,7 @@ func (kd *kubeEngine) CreateDeployment() (watch.Interface, error) {
 	// to have started.
 	w, err := kd.client.CoreV1().Pods(kd.namespace).Watch(opts)
 	if err != nil {
-		return nil, xerrors.Errorf("Pods could not be watched: %w", err)
+		return nil, xerrors.Errorf("pods could not be watched: %v", err)
 	}
 
 	for _, node := range kd.options.Topology.GetNodes() {
@@ -244,7 +244,7 @@ func (kd *kubeEngine) CreateDeployment() (watch.Interface, error) {
 		_, err = kd.client.AppsV1().Deployments(kd.namespace).Create(deployment)
 		if err != nil {
 			fmt.Fprintln(kd.writer, goterm.ResetLine("Creating deployment... failure"))
-			return nil, xerrors.Errorf("Deployment creation failed: %w", err)
+			return nil, xerrors.Errorf("deployment creation failed: %v", err)
 		}
 	}
 
@@ -266,7 +266,7 @@ func (kd *kubeEngine) WaitDeployment(w watch.Interface) error {
 		isReady, err := checkPodStatus(pod)
 		if err != nil {
 			fmt.Fprintln(kd.writer, goterm.ResetLine("Waiting deployment... failure"))
-			return xerrors.Errorf("check pod status failed: %w", err)
+			return xerrors.Errorf("check pod status failed: %v", err)
 		}
 
 		if isReady {
@@ -327,11 +327,11 @@ func (kd *kubeEngine) FetchPods() ([]apiv1.Pod, error) {
 		LabelSelector: fmt.Sprintf("%s=%s", LabelID, AppID),
 	})
 	if err != nil {
-		return nil, xerrors.Errorf("error when listing pods: %w", err)
+		return nil, xerrors.Errorf("error when listing pods: %v", err)
 	}
 
 	if len(pods.Items) != kd.options.Topology.Len() {
-		return nil, fmt.Errorf("invalid number of pods: %d vs %d", len(pods.Items), kd.options.Topology.Len())
+		return nil, xerrors.Errorf("invalid number of pods: %d vs %d", len(pods.Items), kd.options.Topology.Len())
 	}
 
 	fmt.Fprintln(kd.writer, goterm.ResetLine("Fetching pods... ok"))
@@ -415,12 +415,12 @@ func (kd *kubeEngine) DeployRouter(pods []apiv1.Pod) (watch.Interface, error) {
 
 	w, err := kd.client.CoreV1().Pods(kd.namespace).Watch(opts)
 	if err != nil {
-		return nil, xerrors.Errorf("failed watching pods: %w", err)
+		return nil, xerrors.Errorf("failed watching pods: %v", err)
 	}
 
 	_, err = kd.client.AppsV1().Deployments(kd.namespace).Create(makeRouterDeployment())
 	if err != nil {
-		return nil, xerrors.Errorf("failed creating deployment: %w", err)
+		return nil, xerrors.Errorf("failed creating deployment: %v", err)
 	}
 
 	fmt.Fprintln(kd.writer, goterm.ResetLine("Deploying the router... ok"))
@@ -432,12 +432,12 @@ func (kd *kubeEngine) createVPNService() (*apiv1.ServicePort, error) {
 
 	w, err := intf.Watch(metav1.ListOptions{})
 	if err != nil {
-		return nil, xerrors.Errorf("failed watching svc interface: %w", err)
+		return nil, xerrors.Errorf("failed watching svc interface: %v", err)
 	}
 
 	_, err = intf.Create(kd.makeRouterService())
 	if err != nil {
-		return nil, xerrors.Errorf("failed creating router service: %w", err)
+		return nil, xerrors.Errorf("failed creating router service: %v", err)
 	}
 
 	for {
@@ -484,7 +484,7 @@ func (kd *kubeEngine) WaitRouter(w watch.Interface) (*apiv1.ServicePort, string,
 			port, err := kd.createVPNService()
 			if err != nil {
 				return nil, "", xerrors.Errorf(
-					"failed creating vpn: %w", err)
+					"failed creating vpn: %v", err)
 			}
 
 			return port, host, nil
@@ -505,20 +505,20 @@ func (kd *kubeEngine) writeCertificates(pod string, certs sim.Certificates) erro
 	for out, dst := range files {
 		r, err := kd.kio.Read(pod, ContainerRouterName, dst)
 		if err != nil {
-			return xerrors.Errorf("failed reading pod: %w", err)
+			return xerrors.Errorf("failed reading pod: %v", err)
 		}
 
 		// Create a file that can be read only by the user.
 		f, err := os.OpenFile(out, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0600)
 		if err != nil {
-			return xerrors.Errorf("failed opening file: %w", err)
+			return xerrors.Errorf("failed opening file: %v", err)
 		}
 
 		defer f.Close()
 
 		_, err = io.Copy(f, r)
 		if err != nil {
-			return xerrors.Errorf("failed copying file: %w", err)
+			return xerrors.Errorf("failed copying file: %v", err)
 		}
 	}
 
@@ -601,7 +601,7 @@ func (kd *kubeEngine) deleteService(opts *metav1.DeleteOptions) error {
 
 	err := intf.Delete("simnet-router", opts)
 	if err != nil {
-		return xerrors.Errorf("failed deleting router: %w", err)
+		return xerrors.Errorf("failed deleting router: %v", err)
 	}
 
 	return nil
@@ -651,12 +651,12 @@ func (kd *kubeEngine) StreamLogs(ctx context.Context) error {
 	// Clean the folder to remove old log files.
 	err := os.RemoveAll(dir)
 	if err != nil {
-		return fmt.Errorf("couldn't clean log folder: %v", err)
+		return xerrors.Errorf("couldn't clean log folder: %v", err)
 	}
 
 	err = os.MkdirAll(dir, 0755)
 	if err != nil {
-		return fmt.Errorf("couldn't create log folder: %v", err)
+		return xerrors.Errorf("couldn't create log folder: %v", err)
 	}
 
 	kd.wgLogs = sync.WaitGroup{}
@@ -666,7 +666,7 @@ func (kd *kubeEngine) StreamLogs(ctx context.Context) error {
 
 		file, err := os.Create(filepath.Join(dir, fmt.Sprintf("%s.log", pod.Name)))
 		if err != nil {
-			return fmt.Errorf("couldn't create log file: %v", err)
+			return xerrors.Errorf("couldn't create log file: %v", err)
 		}
 
 		req := kd.client.CoreV1().Pods(kd.namespace).GetLogs(pod.Name, &apiv1.PodLogOptions{
@@ -676,7 +676,7 @@ func (kd *kubeEngine) StreamLogs(ctx context.Context) error {
 
 		reader, err := req.Stream()
 		if err != nil {
-			return fmt.Errorf("couldn't open the stream: %w", err)
+			return xerrors.Errorf("couldn't open the stream: %v", err)
 		}
 
 		// This Go routine will be done when the stream is closed.
@@ -700,7 +700,7 @@ func (kd *kubeEngine) ReadStats(pod string, start, end time.Time) (metrics.NodeS
 	reader, err := kd.kio.Read(pod, ContainerMonitorName, MonitorDataFilepath)
 	if err != nil {
 		return metrics.NodeStats{},
-			xerrors.Errorf("failed reading stats: %w", err)
+			xerrors.Errorf("failed reading stats: %v", err)
 	}
 
 	ns := metrics.NewNodeStats(reader, start, end)
@@ -751,7 +751,7 @@ func (kd *kubeEngine) Write(node, path string, content io.Reader) error {
 
 	err := kd.kio.Write(pod.Name, ContainerAppName, path, reader)
 	if err != nil {
-		return fmt.Errorf("couldn't open stream: %w", err)
+		return xerrors.Errorf("couldn't open stream: %v", err)
 	}
 
 	return nil
@@ -767,7 +767,7 @@ func (kd *kubeEngine) Exec(node string, cmd []string, options sim.ExecOptions) e
 
 	err := kd.kio.Exec(pod.Name, ContainerAppName, cmd, options)
 	if err != nil {
-		return fmt.Errorf("couldn't open stream: %w", err)
+		return xerrors.Errorf("couldn't open stream: %v", err)
 	}
 
 	return nil
@@ -831,7 +831,7 @@ func (kd *kubeEngine) FetchStats(start, end time.Time, filename string) error {
 	for _, pod := range kd.pods {
 		ns, err := kd.ReadStats(pod.Name, start, end)
 		if err != nil {
-			return xerrors.Errorf("failed reading stats from pod '%v': %w",
+			return xerrors.Errorf("failed reading stats from pod '%v': %v",
 				pod.Name, err)
 		}
 
@@ -842,7 +842,7 @@ func (kd *kubeEngine) FetchStats(start, end time.Time, filename string) error {
 
 	f, err := os.Create(filename)
 	if err != nil {
-		return xerrors.Errorf("failed creating stats file: %w", err)
+		return xerrors.Errorf("failed creating stats file: %v", err)
 	}
 
 	defer f.Close()
@@ -850,7 +850,7 @@ func (kd *kubeEngine) FetchStats(start, end time.Time, filename string) error {
 	enc := json.NewEncoder(f)
 	err = enc.Encode(&stats)
 	if err != nil {
-		return xerrors.Errorf("failed encoding stats: %w", err)
+		return xerrors.Errorf("failed encoding stats: %v", err)
 	}
 
 	return nil

--- a/sim/kubernetes/engine.go
+++ b/sim/kubernetes/engine.go
@@ -221,7 +221,7 @@ func (kd *kubeEngine) makeContainer() apiv1.Container {
 }
 
 func (kd *kubeEngine) CreateDeployment() (watch.Interface, error) {
-	fmt.Fprint(kd.writer, "Creating deployment...")
+	fmt.Fprintln(kd.writer, "Creating deployment...")
 
 	opts := metav1.ListOptions{
 		LabelSelector: fmt.Sprintf("%s=%s", LabelID, AppID),
@@ -231,7 +231,7 @@ func (kd *kubeEngine) CreateDeployment() (watch.Interface, error) {
 	// to have started.
 	w, err := kd.client.CoreV1().Pods(kd.namespace).Watch(opts)
 	if err != nil {
-		return nil, err
+		return nil, xerrors.Errorf("Pods could not be watched: %w", err)
 	}
 
 	for _, node := range kd.options.Topology.GetNodes() {
@@ -244,7 +244,7 @@ func (kd *kubeEngine) CreateDeployment() (watch.Interface, error) {
 		_, err = kd.client.AppsV1().Deployments(kd.namespace).Create(deployment)
 		if err != nil {
 			fmt.Fprintln(kd.writer, goterm.ResetLine("Creating deployment... failure"))
-			return nil, err
+			return nil, xerrors.Errorf("Deployment creation failed: %w", err)
 		}
 	}
 
@@ -254,7 +254,7 @@ func (kd *kubeEngine) CreateDeployment() (watch.Interface, error) {
 }
 
 func (kd *kubeEngine) WaitDeployment(w watch.Interface) error {
-	fmt.Fprint(kd.writer, "Waiting deployment...")
+	fmt.Fprintln(kd.writer, "Waiting deployment...")
 	readyMap := make(map[string]struct{})
 
 	for {
@@ -266,7 +266,7 @@ func (kd *kubeEngine) WaitDeployment(w watch.Interface) error {
 		isReady, err := checkPodStatus(pod)
 		if err != nil {
 			fmt.Fprintln(kd.writer, goterm.ResetLine("Waiting deployment... failure"))
-			return err
+			return xerrors.Errorf("check pod status failed: %w", err)
 		}
 
 		if isReady {
@@ -294,7 +294,7 @@ func (kd *kubeEngine) configureContainer(pods []apiv1.Pod) error {
 
 	errCh := make(chan error, len(pods))
 
-	fmt.Fprintf(kd.writer, "Configuring pods...")
+	fmt.Fprintln(kd.writer, "Configuring pods...")
 	for _, pod := range pods {
 		go func(pod apiv1.Pod) {
 			defer wg.Done()
@@ -321,13 +321,13 @@ func (kd *kubeEngine) configureContainer(pods []apiv1.Pod) error {
 }
 
 func (kd *kubeEngine) FetchPods() ([]apiv1.Pod, error) {
-	fmt.Fprintf(kd.writer, "Fetching pods...")
+	fmt.Fprintln(kd.writer, "Fetching pods...")
 
 	pods, err := kd.client.CoreV1().Pods(kd.namespace).List(metav1.ListOptions{
 		LabelSelector: fmt.Sprintf("%s=%s", LabelID, AppID),
 	})
 	if err != nil {
-		return nil, err
+		return nil, xerrors.Errorf("error when listing pods: %w", err)
 	}
 
 	if len(pods.Items) != kd.options.Topology.Len() {
@@ -407,7 +407,7 @@ func (kd *kubeEngine) UploadConfig() error {
 }
 
 func (kd *kubeEngine) DeployRouter(pods []apiv1.Pod) (watch.Interface, error) {
-	fmt.Fprintf(kd.writer, "Deploying the router...")
+	fmt.Fprintln(kd.writer, "Deploying the router...")
 
 	opts := metav1.ListOptions{
 		LabelSelector: fmt.Sprintf("%s=%s", LabelID, RouterID),
@@ -415,12 +415,12 @@ func (kd *kubeEngine) DeployRouter(pods []apiv1.Pod) (watch.Interface, error) {
 
 	w, err := kd.client.CoreV1().Pods(kd.namespace).Watch(opts)
 	if err != nil {
-		return nil, err
+		return nil, xerrors.Errorf("failed watching pods: %w", err)
 	}
 
 	_, err = kd.client.AppsV1().Deployments(kd.namespace).Create(makeRouterDeployment())
 	if err != nil {
-		return nil, err
+		return nil, xerrors.Errorf("failed creating deployment: %w", err)
 	}
 
 	fmt.Fprintln(kd.writer, goterm.ResetLine("Deploying the router... ok"))
@@ -432,12 +432,12 @@ func (kd *kubeEngine) createVPNService() (*apiv1.ServicePort, error) {
 
 	w, err := intf.Watch(metav1.ListOptions{})
 	if err != nil {
-		return nil, err
+		return nil, xerrors.Errorf("failed watching svc interface: %w", err)
 	}
 
 	_, err = intf.Create(kd.makeRouterService())
 	if err != nil {
-		return nil, err
+		return nil, xerrors.Errorf("failed creating router service: %w", err)
 	}
 
 	for {
@@ -453,7 +453,7 @@ func (kd *kubeEngine) createVPNService() (*apiv1.ServicePort, error) {
 }
 
 func (kd *kubeEngine) WaitRouter(w watch.Interface) (*apiv1.ServicePort, string, error) {
-	fmt.Fprintf(kd.writer, "Waiting for the router...")
+	fmt.Fprintln(kd.writer, "Waiting for the router...")
 
 	host := kd.host
 
@@ -483,7 +483,8 @@ func (kd *kubeEngine) WaitRouter(w watch.Interface) (*apiv1.ServicePort, string,
 			fmt.Fprintln(kd.writer, goterm.ResetLine("Waiting for the router... ok"))
 			port, err := kd.createVPNService()
 			if err != nil {
-				return nil, "", err
+				return nil, "", xerrors.Errorf(
+					"failed creating vpn: %w", err)
 			}
 
 			return port, host, nil
@@ -504,20 +505,20 @@ func (kd *kubeEngine) writeCertificates(pod string, certs sim.Certificates) erro
 	for out, dst := range files {
 		r, err := kd.kio.Read(pod, ContainerRouterName, dst)
 		if err != nil {
-			return err
+			return xerrors.Errorf("failed reading pod: %w", err)
 		}
 
 		// Create a file that can be read only by the user.
 		f, err := os.OpenFile(out, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0600)
 		if err != nil {
-			return err
+			return xerrors.Errorf("failed opening file: %w", err)
 		}
 
 		defer f.Close()
 
 		_, err = io.Copy(f, r)
 		if err != nil {
-			return err
+			return xerrors.Errorf("failed copying file: %w", err)
 		}
 	}
 
@@ -525,7 +526,7 @@ func (kd *kubeEngine) writeCertificates(pod string, certs sim.Certificates) erro
 }
 
 func (kd *kubeEngine) FetchCertificates() (sim.Certificates, error) {
-	fmt.Fprintf(kd.writer, "Fetching the certificates...")
+	fmt.Fprintln(kd.writer, "Fetching the certificates...")
 
 	// This defines where the files will be written on the client side.
 	certs := sim.Certificates{
@@ -600,7 +601,7 @@ func (kd *kubeEngine) deleteService(opts *metav1.DeleteOptions) error {
 
 	err := intf.Delete("simnet-router", opts)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed deleting router: %w", err)
 	}
 
 	return nil
@@ -698,7 +699,8 @@ func (kd *kubeEngine) StreamLogs(ctx context.Context) error {
 func (kd *kubeEngine) ReadStats(pod string, start, end time.Time) (metrics.NodeStats, error) {
 	reader, err := kd.kio.Read(pod, ContainerMonitorName, MonitorDataFilepath)
 	if err != nil {
-		return metrics.NodeStats{}, err
+		return metrics.NodeStats{},
+			xerrors.Errorf("failed reading stats: %w", err)
 	}
 
 	ns := metrics.NewNodeStats(reader, start, end)
@@ -829,7 +831,8 @@ func (kd *kubeEngine) FetchStats(start, end time.Time, filename string) error {
 	for _, pod := range kd.pods {
 		ns, err := kd.ReadStats(pod.Name, start, end)
 		if err != nil {
-			return err
+			return xerrors.Errorf("failed reading stats from pod '%v': %w",
+				pod.Name, err)
 		}
 
 		name := pod.Labels[LabelNode]
@@ -839,7 +842,7 @@ func (kd *kubeEngine) FetchStats(start, end time.Time, filename string) error {
 
 	f, err := os.Create(filename)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed creating stats file: %w", err)
 	}
 
 	defer f.Close()
@@ -847,7 +850,7 @@ func (kd *kubeEngine) FetchStats(start, end time.Time, filename string) error {
 	enc := json.NewEncoder(f)
 	err = enc.Encode(&stats)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed encoding stats: %w", err)
 	}
 
 	return nil

--- a/sim/kubernetes/io.go
+++ b/sim/kubernetes/io.go
@@ -112,7 +112,7 @@ func (k kio) Exec(pod, container string, cmd []string, options sim.ExecOptions) 
 
 	exec, err := newExecutor(k.config, "POST", req.URL())
 	if err != nil {
-		return xerrors.Errorf("failed executing cmd '%v' on pod: %w",
+		return xerrors.Errorf("failed executing cmd '%v' on pod: %v",
 			cmd[0], err)
 	}
 

--- a/sim/kubernetes/io.go
+++ b/sim/kubernetes/io.go
@@ -95,6 +95,10 @@ func (k kio) Write(pod, container, path string, content io.Reader) error {
 }
 
 func (k kio) Exec(pod, container string, cmd []string, options sim.ExecOptions) error {
+	if len(cmd) < 1 {
+		return xerrors.New("no command to execute")
+	}
+
 	req := k.restclient.
 		Post().
 		Namespace(k.namespace).
@@ -112,7 +116,7 @@ func (k kio) Exec(pod, container string, cmd []string, options sim.ExecOptions) 
 
 	exec, err := newExecutor(k.config, "POST", req.URL())
 	if err != nil {
-		return xerrors.Errorf("failed executing cmd '%v' on pod: %v",
+		return xerrors.Errorf("failed executing cmd '%s' on pod: %v",
 			cmd[0], err)
 	}
 

--- a/sim/kubernetes/io.go
+++ b/sim/kubernetes/io.go
@@ -112,7 +112,8 @@ func (k kio) Exec(pod, container string, cmd []string, options sim.ExecOptions) 
 
 	exec, err := newExecutor(k.config, "POST", req.URL())
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed executing cmd '%v' on pod: %w",
+			cmd[0], err)
 	}
 
 	done := make(chan error)

--- a/sim/kubernetes/mod.go
+++ b/sim/kubernetes/mod.go
@@ -115,46 +115,46 @@ func (s *Strategy) Option(opt sim.Option) {
 func (s *Strategy) Deploy(ctx context.Context, round sim.Round) error {
 	w, err := s.engine.CreateDeployment()
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed creating deployment: %w", err)
 	}
 
 	err = s.engine.WaitDeployment(w)
 	w.Stop()
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed waiting deployment: %w", err)
 	}
 
 	pods, err := s.engine.FetchPods()
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed fetching pods: %w", err)
 	}
 
 	err = s.engine.StreamLogs(ctx)
 	if err != nil {
-		return xerrors.Errorf("couldn't stream logs: %v", err)
+		return xerrors.Errorf("failed streaming logs: %v", err)
 	}
 
 	err = s.engine.UploadConfig()
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed uploading config: %w", err)
 	}
 
 	s.pods = pods
 
 	w, err = s.engine.DeployRouter(pods)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed deploying router: %w", err)
 	}
 
 	port, host, err := s.engine.WaitRouter(w)
 	w.Stop()
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed waiting router: %w", err)
 	}
 
 	certificates, err := s.engine.FetchCertificates()
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed fetching certs: %w", err)
 	}
 
 	err = s.tun.Start(
@@ -164,14 +164,14 @@ func (s *Strategy) Deploy(ctx context.Context, round sim.Round) error {
 		sim.WithCommand(s.options.VPNExecutable),
 	)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed starting tunnel: %w", err)
 	}
 
 	// Before is run at the end of the deployment so that the Execute
 	// step can be run multiple times.
 	err = round.Before(s.engine, s.makeContext())
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed running 'Before': %w", err)
 	}
 
 	s.updated = true

--- a/sim/kubernetes/mod.go
+++ b/sim/kubernetes/mod.go
@@ -73,7 +73,7 @@ func NewStrategy(cfg string, opts ...sim.Option) (*Strategy, error) {
 	// Create the directory where the data will be stored.
 	err := os.MkdirAll(options.OutputDir, 0755)
 	if err != nil {
-		return nil, fmt.Errorf("couldn't create the data folder: %v", err)
+		return nil, xerrors.Errorf("couldn't create the data folder: %v", err)
 	}
 
 	clientcfg := newClientConfig(
@@ -83,17 +83,17 @@ func NewStrategy(cfg string, opts ...sim.Option) (*Strategy, error) {
 
 	restcfg, err := clientcfg.ClientConfig()
 	if err != nil {
-		return nil, fmt.Errorf("config: %v", err)
+		return nil, xerrors.Errorf("config: %v", err)
 	}
 
 	namespace, _, err := clientcfg.Namespace()
 	if err != nil {
-		return nil, fmt.Errorf("couldn't read the namespace: %v", err)
+		return nil, xerrors.Errorf("couldn't read the namespace: %v", err)
 	}
 
 	engine, err := newKubeEngine(restcfg, namespace, options)
 	if err != nil {
-		return nil, fmt.Errorf("couldn't create the engine: %v", err)
+		return nil, xerrors.Errorf("couldn't create the engine: %v", err)
 	}
 
 	return &Strategy{
@@ -115,18 +115,18 @@ func (s *Strategy) Option(opt sim.Option) {
 func (s *Strategy) Deploy(ctx context.Context, round sim.Round) error {
 	w, err := s.engine.CreateDeployment()
 	if err != nil {
-		return xerrors.Errorf("failed creating deployment: %w", err)
+		return xerrors.Errorf("failed creating deployment: %v", err)
 	}
 
 	err = s.engine.WaitDeployment(w)
 	w.Stop()
 	if err != nil {
-		return xerrors.Errorf("failed waiting deployment: %w", err)
+		return xerrors.Errorf("failed waiting deployment: %v", err)
 	}
 
 	pods, err := s.engine.FetchPods()
 	if err != nil {
-		return xerrors.Errorf("failed fetching pods: %w", err)
+		return xerrors.Errorf("failed fetching pods: %v", err)
 	}
 
 	err = s.engine.StreamLogs(ctx)
@@ -136,25 +136,25 @@ func (s *Strategy) Deploy(ctx context.Context, round sim.Round) error {
 
 	err = s.engine.UploadConfig()
 	if err != nil {
-		return xerrors.Errorf("failed uploading config: %w", err)
+		return xerrors.Errorf("failed uploading config: %v", err)
 	}
 
 	s.pods = pods
 
 	w, err = s.engine.DeployRouter(pods)
 	if err != nil {
-		return xerrors.Errorf("failed deploying router: %w", err)
+		return xerrors.Errorf("failed deploying router: %v", err)
 	}
 
 	port, host, err := s.engine.WaitRouter(w)
 	w.Stop()
 	if err != nil {
-		return xerrors.Errorf("failed waiting router: %w", err)
+		return xerrors.Errorf("failed waiting router: %v", err)
 	}
 
 	certificates, err := s.engine.FetchCertificates()
 	if err != nil {
-		return xerrors.Errorf("failed fetching certs: %w", err)
+		return xerrors.Errorf("failed fetching certs: %v", err)
 	}
 
 	err = s.tun.Start(
@@ -164,14 +164,14 @@ func (s *Strategy) Deploy(ctx context.Context, round sim.Round) error {
 		sim.WithCommand(s.options.VPNExecutable),
 	)
 	if err != nil {
-		return xerrors.Errorf("failed starting tunnel: %w", err)
+		return xerrors.Errorf("failed starting tunnel: %v", err)
 	}
 
 	// Before is run at the end of the deployment so that the Execute
 	// step can be run multiple times.
 	err = round.Before(s.engine, s.makeContext())
 	if err != nil {
-		return xerrors.Errorf("failed running 'Before': %w", err)
+		return xerrors.Errorf("failed running 'Before': %v", err)
 	}
 
 	s.updated = true
@@ -210,14 +210,14 @@ func (s *Strategy) Execute(ctx context.Context, round sim.Round) error {
 
 	err = round.Execute(s.engine, nodes)
 	if err != nil {
-		return xerrors.Errorf("couldn't perform execute step: %w", err)
+		return xerrors.Errorf("couldn't perform execute step: %v", err)
 	}
 
 	s.doneTime = time.Now()
 
 	err = round.After(s.engine, nodes)
 	if err != nil {
-		return xerrors.Errorf("couldn't perform after step: %w", err)
+		return xerrors.Errorf("couldn't perform after step: %v", err)
 	}
 
 	return nil
@@ -268,7 +268,7 @@ func (s *Strategy) Clean(ctx context.Context) error {
 	}
 
 	if len(errs) > 0 {
-		return fmt.Errorf("cleaning error: %v", errs)
+		return xerrors.Errorf("cleaning error: %v", errs)
 	}
 
 	return nil

--- a/sim/option.go
+++ b/sim/option.go
@@ -1,7 +1,9 @@
 package sim
 
 import (
+	"errors"
 	"fmt"
+	"io/fs"
 	"os"
 	"path/filepath"
 	"strings"
@@ -67,6 +69,12 @@ func NewOptions(opts []Option) *Options {
 		}
 
 		o.OutputDir = filepath.Join(homeDir, ".config", "simnet")
+	}
+
+	err := os.Mkdir(o.OutputDir, 0755)
+	if err != nil && !errors.Is(err, fs.ErrExist) {
+		fmt.Println("failed creating simnet config dir: ", o.OutputDir, err)
+		panic(err)
 	}
 
 	return o

--- a/sim/tunnel.go
+++ b/sim/tunnel.go
@@ -121,14 +121,14 @@ func (v *DefaultTunnel) Start(opts ...TunOption) error {
 
 	file, err := os.Create(filepath.Join(v.outDir, LogFileName))
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed creating VPN log: %w", err)
 	}
 
 	defer file.Close()
 
 	usr, err := currentUser()
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed getting current user: %w", err)
 	}
 
 	args := []string{
@@ -209,7 +209,7 @@ func (v *DefaultTunnel) Stop() error {
 		if err == nil {
 			proc, err := findProcess(pid)
 			if err != nil {
-				return err
+				return xerrors.Errorf("failed to find process: %w", err)
 			}
 
 			err = proc.Kill()
@@ -217,7 +217,7 @@ func (v *DefaultTunnel) Stop() error {
 				if err.Error() == "os: process already finished" {
 					return nil
 				}
-				return err
+				return xerrors.Errorf("failed to kill: %w", err)
 			}
 		}
 	}

--- a/sim/tunnel.go
+++ b/sim/tunnel.go
@@ -121,14 +121,14 @@ func (v *DefaultTunnel) Start(opts ...TunOption) error {
 
 	file, err := os.Create(filepath.Join(v.outDir, LogFileName))
 	if err != nil {
-		return xerrors.Errorf("failed creating VPN log: %w", err)
+		return xerrors.Errorf("failed creating VPN log: %v", err)
 	}
 
 	defer file.Close()
 
 	usr, err := currentUser()
 	if err != nil {
-		return xerrors.Errorf("failed getting current user: %w", err)
+		return xerrors.Errorf("failed getting current user: %v", err)
 	}
 
 	args := []string{
@@ -172,7 +172,7 @@ func (v *DefaultTunnel) Start(opts ...TunOption) error {
 	err = cmd.Run()
 	if err != nil {
 		fmt.Printf("Command 'sudo %s' returns error: %s\n", options.Cmd, err)
-		return fmt.Errorf("vpn initialization failed: see %s", filepath.Join(v.outDir, LogFileName))
+		return xerrors.Errorf("vpn initialization failed: see %s", filepath.Join(v.outDir, LogFileName))
 	}
 
 	timeout := time.After(v.timeout)
@@ -209,7 +209,7 @@ func (v *DefaultTunnel) Stop() error {
 		if err == nil {
 			proc, err := findProcess(pid)
 			if err != nil {
-				return xerrors.Errorf("failed to find process: %w", err)
+				return xerrors.Errorf("failed to find process: %v", err)
 			}
 
 			err = proc.Kill()
@@ -217,7 +217,7 @@ func (v *DefaultTunnel) Stop() error {
 				if err.Error() == "os: process already finished" {
 					return nil
 				}
-				return xerrors.Errorf("failed to kill: %w", err)
+				return xerrors.Errorf("failed to kill: %v", err)
 			}
 		}
 	}

